<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference • syntenet</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.6/font.css" rel="stylesheet"><link href="../deps/Source_Code_Pro-0.4.6/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-#0092ac"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">syntenet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/syntenet.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/almeidasilvaf/syntenet/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="angiosperm_phylogeny.html">angiosperm_phylogeny</a></code> 
        </dt>
        <dd>Microsynteny-based angiosperm phylogeny.</dd>
      </dl><dl><dt>
          
          <code><a href="annotation.html">annotation</a></code> 
        </dt>
        <dd>Filtered genome annotation for Ostreococcus sp. species</dd>
      </dl><dl><dt>
          
          <code><a href="binarize_and_transpose.html">binarize_and_transpose()</a></code> 
        </dt>
        <dd>Binarize and transpose the phylogenomic profile matrix</dd>
      </dl><dl><dt>
          
          <code><a href="blast_list.html">blast_list</a></code> 
        </dt>
        <dd>List of data frames containing BLAST-like tabular output</dd>
      </dl><dl><dt>
          
          <code><a href="check_input.html">check_input()</a></code> 
        </dt>
        <dd>Check if input objects are ready for further analyses</dd>
      </dl><dl><dt>
          
          <code><a href="cluster_network.html">cluster_network()</a></code> 
        </dt>
        <dd>Cluster the synteny network using the Infomap algorithm</dd>
      </dl><dl><dt>
          
          <code><a href="clusters.html">clusters</a></code> 
        </dt>
        <dd>Synteny network clusters of BUSCO genes for 25 eudicot species</dd>
      </dl><dl><dt>
          
          <code><a href="collapse_protein_ids.html">collapse_protein_ids()</a></code> 
        </dt>
        <dd>Collapse protein IDs into gene IDs in sequence names of AAStringSet objects</dd>
      </dl><dl><dt>
          
          <code><a href="create_species_id_table.html">create_species_id_table()</a></code> 
        </dt>
        <dd>Create a data frame of species IDs (3-5-character abbreviations)</dd>
      </dl><dl><dt>
          
          <code><a href="diamond_is_installed.html">diamond_is_installed()</a></code> 
        </dt>
        <dd>Check if DIAMOND is installed</dd>
      </dl><dl><dt>
          
          <code><a href="edges.html">edges</a></code> 
        </dt>
        <dd>Synteny network of Ostreococcus genomes represented as an edge list</dd>
      </dl><dl><dt>
          
          <code><a href="export_sequences.html">export_sequences()</a></code> 
        </dt>
        <dd>Export processed sequences as FASTA files</dd>
      </dl><dl><dt>
          
          <code><a href="fasta2AAStringSetlist.html">fasta2AAStringSetlist()</a></code> 
        </dt>
        <dd>Read FASTA files in a directory as a list of AAStringSet objects</dd>
      </dl><dl><dt>
          
          <code><a href="find_GS_clusters.html">find_GS_clusters()</a></code> 
        </dt>
        <dd>Find group-specific clusters based on user-defined species classification</dd>
      </dl><dl><dt>
          
          <code><a href="gff2GRangesList.html">gff2GRangesList()</a></code> 
        </dt>
        <dd>Read GFF/GTF files in a directory as a GRangesList object</dd>
      </dl><dl><dt>
          
          <code><a href="infer_microsynteny_phylogeny.html">infer_microsynteny_phylogeny()</a></code> 
        </dt>
        <dd>Infer microsynteny-based phylogeny with IQTREE</dd>
      </dl><dl><dt>
          
          <code><a href="infer_syntenet.html">infer_syntenet()</a></code> 
        </dt>
        <dd>Infer synteny network</dd>
      </dl><dl><dt>
          
          <code><a href="interspecies_synteny.html">interspecies_synteny()</a></code> 
        </dt>
        <dd>Detect interspecies synteny</dd>
      </dl><dl><dt>
          
          <code><a href="intraspecies_synteny.html">intraspecies_synteny()</a></code> 
        </dt>
        <dd>Detect intraspecies synteny</dd>
      </dl><dl><dt>
          
          <code><a href="iqtree_is_installed.html">iqtree_is_installed()</a></code> 
        </dt>
        <dd>Check if IQTREE is installed</dd>
      </dl><dl><dt>
          
          <code><a href="iqtree_version.html">iqtree_version()</a></code> 
        </dt>
        <dd>Get IQ-TREE version</dd>
      </dl><dl><dt>
          
          <code><a href="network.html">network</a></code> 
        </dt>
        <dd>Synteny network of BUSCO genes for 25 eudicot species</dd>
      </dl><dl><dt>
          
          <code><a href="parse_collinearity.html">parse_collinearity()</a></code> 
        </dt>
        <dd>Parse .collinearity files obtained with MCScan</dd>
      </dl><dl><dt>
          
          <code><a href="phylogenomic_profile.html">phylogenomic_profile()</a></code> 
        </dt>
        <dd>Perform phylogenomic profiling for synteny network clusters</dd>
      </dl><dl><dt>
          
          <code><a href="plot_network.html">plot_network()</a></code> 
        </dt>
        <dd>Plot network</dd>
      </dl><dl><dt>
          
          <code><a href="plot_profiles.html">plot_profiles()</a></code> 
        </dt>
        <dd>Plot a heatmap of phylogenomic profiles</dd>
      </dl><dl><dt>
          
          <code><a href="process_input.html">process_input()</a></code> 
        </dt>
        <dd>Process sequence data</dd>
      </dl><dl><dt>
          
          <code><a href="profiles2phylip.html">profiles2phylip()</a></code> 
        </dt>
        <dd>Save the transposed binary profiles matrix to a file in PHYLIP format</dd>
      </dl><dl><dt>
          
          <code><a href="proteomes.html">proteomes</a></code> 
        </dt>
        <dd>Filtered proteomes of Ostreococcus sp. species</dd>
      </dl><dl><dt>
          
          <code><a href="rcpp_mcscanx_file.html">rcpp_mcscanx_file()</a></code> 
        </dt>
        <dd>rcpp_mcscanx_file</dd>
      </dl><dl><dt>
          
          <code><a href="read_diamond.html">read_diamond()</a></code> 
        </dt>
        <dd>Read DIAMOND/BLAST tables as a list of data frames</dd>
      </dl><dl><dt>
          
          <code><a href="run_diamond.html">run_diamond()</a></code> 
        </dt>
        <dd>Wrapper to run DIAMOND from an R session</dd>
      </dl><dl><dt>
          
          <code><a href="scerevisiae_annot.html">scerevisiae_annot</a></code> 
        </dt>
        <dd>Genome annotation of the yeast species S. cerevisiae</dd>
      </dl><dl><dt>
          
          <code><a href="scerevisiae_diamond.html">scerevisiae_diamond</a></code> 
        </dt>
        <dd>Intraspecies DIAMOND output for S. cerevisiae</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Fabrício Almeida-Silva, Tao Zhao, Kristian K Ullrich, Yves Van de Peer.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

