<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>rcpp_mcscanx_file — rcpp_mcscanx_file • syntenet</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="rcpp_mcscanx_file — rcpp_mcscanx_file"><meta property="og:description" content="MCSanX provides a clustering module for viewing the
relationship of colinear segments in multiple genomes (or heavily redundant
genomes). It takes the predicted pairwise segments from dynamic programming
(DAGchainer in particular) and then try to build consensus segments from a
set of related, overlapping segments."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">syntenet</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/syntenet.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/almeidasilvaf/syntenet/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>rcpp_mcscanx_file</h1>
    <small class="dont-index">Source: <a href="https://github.com/almeidasilvaf/syntenet/blob/HEAD/R/RcppExports.R" class="external-link"><code>R/RcppExports.R</code></a></small>
    <div class="hidden name"><code>rcpp_mcscanx_file.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>MCSanX provides a clustering module for viewing the
relationship of colinear segments in multiple genomes (or heavily redundant
genomes). It takes the predicted pairwise segments from dynamic programming
(DAGchainer in particular) and then try to build consensus segments from a
set of related, overlapping segments.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">rcpp_mcscanx_file</span><span class="op">(</span></span>
<span>  <span class="va">blast_file</span>,</span>
<span>  <span class="va">gff_file</span>,</span>
<span>  prefix <span class="op">=</span> <span class="st">"out"</span>,</span>
<span>  outdir <span class="op">=</span> <span class="st">""</span>,</span>
<span>  match_score <span class="op">=</span> <span class="fl">50L</span>,</span>
<span>  gap_penalty <span class="op">=</span> <span class="op">-</span><span class="fl">1L</span>,</span>
<span>  match_size <span class="op">=</span> <span class="fl">5L</span>,</span>
<span>  e_value <span class="op">=</span> <span class="fl">1e-05</span>,</span>
<span>  max_gaps <span class="op">=</span> <span class="fl">25L</span>,</span>
<span>  overlap_window <span class="op">=</span> <span class="fl">5L</span>,</span>
<span>  is_pairwise <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  in_synteny <span class="op">=</span> <span class="fl">0L</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>blast_file</dt>
<dd><p>blast input</p></dd>


<dt>gff_file</dt>
<dd><p>gff input</p></dd>


<dt>prefix</dt>
<dd><p>output prefix (default: out)</p></dd>


<dt>outdir</dt>
<dd><p>output directory (default: "")</p></dd>


<dt>match_score</dt>
<dd><p>match score (default: 50)</p></dd>


<dt>gap_penalty</dt>
<dd><p>gap penalty (default: -1)</p></dd>


<dt>match_size</dt>
<dd><p>match_size (default: 5)</p></dd>


<dt>e_value</dt>
<dd><p>e_value (default: 1e-5)</p></dd>


<dt>max_gaps</dt>
<dd><p>max gaps (default: 25)</p></dd>


<dt>overlap_window</dt>
<dd><p>overlap window (default: 5)</p></dd>


<dt>is_pairwise</dt>
<dd><p>specify if only pairwise blocks should be reported
(default: FALSE)</p></dd>


<dt>in_synteny</dt>
<dd><p>specify patterns of collinear blocks.
0: intra- and inter-species (default); 1: intra-species; 2: inter-species</p></dd>


<dt>verbose</dt>
<dd><p>specify if verbose output (default: FALSE)</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>list</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Wang et al. (2012) MCScanX: a toolkit for detection and
evolutionary analysis of gene synteny and collinearity.
<em>Nucleic acids research</em>. <b>40.7</b>, e49-e49.</p>
<p>Haas et al. (2004) DAGchainer: a tool for mining segmental
genome duplications and synteny. <em>Bioinformatics</em>. <b>20.18</b>
3643-3646.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Kristian K Ullrich</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Fabrício Almeida-Silva, Tao Zhao, Kristian K Ullrich, Yves Van de Peer.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

